src/pathfinding/grid_pathfinder.jl

```diff
@@ -0,0 +1,362 @@
+export CostMetric,
+    DirectDistance,
+    Chebyshev,
+    HeightMap,
+    AStar,
+    delta_cost,
+    find_path,
+    set_target!,
+    is_stationary,
+    heightmap,
+    walkmap
+
+"""
+    Path{D}
+Alias of `MutableLinkedList{Dims{D}}`. Used to represent the path to be
+taken by an agent in a `D` dimensional [`GridSpace{D}`](@ref).
+"""
+const Path{D} = MutableLinkedList{Dims{D}}
+
+"""
+    CostMetric{D}
+An abstract type representing a metric that measures the approximate cost of travelling
+between two points in a `D` dimensional [`GridSpace{D}`](@ref).
+"""
+abstract type CostMetric{D} end
+
+struct DirectDistance{D} <: CostMetric{D}
+    direction_costs::Vector{Int}
+end
+
+Base.show(io::IO, metric::DirectDistance) = print(io, "DirectDistance")
+
+"""
+    DirectDistance
+Distance is approximated as the shortest path between the two points, provided the
+`walkable` property of [`AStar`](@ref) allows.
+Optionally provide a `Vector{Int}` that represents the cost of going from a tile to the
+neighboring tile on the `i` dimensional diagonal (default is `10√i`).
+
+If `moore_neighbors=false` in [`AStar`](@ref), only Von Neumann neighbors will be tested.
+Cost defaults to the first value of the provided vector.
+"""
+DirectDistance{D}() where {D} = DirectDistance{D}([floor(Int, 10.0 * √x) for x in 1:D])
+
+"""
+    Chebyshev
+Distance between two tiles is approximated as the Chebyshev distance (maximum of absolute
+difference in coordinates) between them.
+"""
+struct Chebyshev{D} <: CostMetric{D} end
+
+Base.show(io::IO, metric::Chebyshev) = print(io, "Chebyshev")
+
+struct HeightMap{D} <: CostMetric{D}
+    base_metric::CostMetric{D}
+    hmap::Array{Int,D}
+end
+
+Base.show(io::IO, metric::HeightMap) =
+    print(io, "HeightMap with base: $(metric.base_metric)")
+
+"""
+    HeightMap(hmap::Array{Int,D})
+    HeightMap(hmap::Array{Int,D}, ::Type{<:CostMetric})
```

Comment:
It is not clear that `HeightMap` is a subtype of `CostMetric`... Overall I think all documentation strings related to `AStar` need to be improved a bit for clarity.

Reply:
I'm not sure what the format is for specifying both the subtype and constructor parameters in the docstring. Could you give an example?

Reply:
`HeightMap(hmap::Array{Int,D}, cm::CostMetric) <: CostMetric`

oh wait, the second argument is not an instance of a type, but a type itself? Why?

Reply:
That's me. Same as the `Pathfinder` struct accepted a type or an instance. Just for ease of use. Better for users to make `HeightMap(hmap)` or `Chebyshev` rather than `HeightMap(hmap)` (which must use this interface to create an instance), or `Chebyshev{2}()` (which requires the dimensional argument to be passed).